### PR TITLE
Refactor and simplify EvmState to encapsulate tracking of warm/cold access

### PR DESCRIPTION
**Pull request by ak88**
Enables warmup of addresses/storage outside `EvmState`, which then contains a tracker per "stack frame".

After review comment, `AccessTracker `is now a struct

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/EvmState.cs [hunk 1]
```diff
@@ -107,8 +107,25 @@ public void ReturnStacks(byte[] dataStack, int[] returnStack)
 
         public int ReturnStackHead = 0;
         private bool _canRestore = true;
-
         public EvmState(
+            long gasAvailable,
+            ExecutionEnvironment env,
+            ExecutionType executionType,
+            bool isTopLevel,
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/EvmState.cs:114]: I think this should be removed as parameter and hardcoded to true, not to use this for not top level states

File: src/Nethermind/Nethermind.Evm/AccessTracker.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+namespace Nethermind.Evm
+{
+    public sealed class AccessTracker
+    {
+        public JournalSet<Address> AccessedAddresses { get; }
+        public JournalSet<StorageCell> AccessedStorageCells { get; }
+        public JournalCollection<LogEntry> Logs { get; }
+        public JournalSet<Address> DestroyList { get; }
+        public HashSet<AddressAsKey> CreateList { get; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/AccessTracker.cs:19]: If they are public should they be `IReadOnlySet`?

File: src/Nethermind/Nethermind.Evm/EvmState.cs [hunk 3]
```diff
@@ -253,28 +251,13 @@ public void InitStacks()
...
-                {
-                    WarmUp(address);
-                    foreach (UInt256 storage in storages)
-                    {
-                        WarmUp(new StorageCell(address, storage));
-                    }
-                }
-            }
-        }
+        public bool IsCold(in StorageCell storageCell) => !_accessTracker.AccessedStorageCells.Contains(storageCell);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/EvmState.cs:256]: Use AccessTracker methods

File: src/Nethermind/Nethermind.Evm/EvmState.cs [hunk 4]
```diff
@@ -162,32 +173,19 @@ internal EvmState(
...
             else
             {
-                // if we are top level, then we need to create the collections
-                _accessedAddresses = new JournalSet<Address>();
-                _accessedStorageCells = new JournalSet<StorageCell>();
-                _destroyList = new JournalSet<Address>();
-                _createList = new HashSet<AddressAsKey>();
-                _logs = new JournalCollection<LogEntry>();
+                // if we are top level, then we create a new access tracker
+                _accessTracker = new();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/EvmState.cs:182]: Wonder if we could/should move this code upper?

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 5]
```diff
@@ -150,13 +150,13 @@ protected virtual TransactionResult Execute(Transaction tx, in BlockExecutionCon
 
             if (commit) WorldState.Commit(spec, tracer.IsTracingState ? tracer : NullTxTracer.Instance, commitStorageRoots: false);
 
-            _accessedAddresses.Clear();
-            int delegationRefunds = ProcessDelegations(tx, spec, _accessedAddresses);
+            AccessTracker accessTracker = new();
+            int delegationRefunds = ProcessDelegations(tx, spec, accessTracker.AccessedAddresses);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:154]: pass whole accessTracker

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 6]
```diff
@@ -507,24 +507,37 @@ private ExecutionEnvironment BuildExecutionEnvironment(
...
+                    accessedAddresses.Add(tx.AccessList); // eip-2930
+                }
+
+                if (spec.AddCoinbaseToTxAccessList)
+                {
+                    accessedAddresses.Add(blCtx.Header.GasBeneficiary!);
+                }
+
+                //We assume eip-7702 must be active if it is a delegation 
+                if (delegationAddress is not null)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:538]: inconsistent parenthesis

**Review by LukaszRozmej — COMMENTED** [review]
Overall cool, but if we are going encapsulation, why not go the full way?
Wonder if we can simplify `AccessTracker` creation. Now we have it complicated between `AccessTracker` constructor and `EvmState` constructor.
Wonder if `AccessTracker` should be a struct? (Then pass it with `in` everywhere)

File: src/Nethermind/Nethermind.Evm/EvmState.cs (hunk 1 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm/EvmState.cs]: removed

File: src/Nethermind/Nethermind.Evm/AccessTracker.cs (hunk 2 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm/AccessTracker.cs]: All made into `IReadOnlySet`

File: src/Nethermind/Nethermind.Evm/EvmState.cs (hunk 3 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm/EvmState.cs]: changed

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs (hunk 5 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs]: changed

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs (hunk 6 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:559]: fixed

File: src/Nethermind/Nethermind.Evm/EvmState.cs (hunk 4 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm/EvmState.cs]: code has been simplified and refactored.

**Comment by ak88** [comment]
> Overall cool, but if we are going encapsulation, why not go the full way? Wonder if we can simplify `AccessTracker` creation. Now we have it complicated between `AccessTracker` constructor and `EvmState` constructor. Wonder if `AccessTracker` should be a struct? (Then pass it with `in` everywhere)

Refactored `EvmState` ctors to be simpler and made `AccessTracker` a struct.

**Review by LukaszRozmej — APPROVED** [review]
@benaadams @Demuirgos  can you take a look?

**Review by benaadams — APPROVED** [review]

**Review by Demuirgos — APPROVED** [review]

**Pull request merged**